feat: initial implementation [BB-7595]

Agrendalath opened the pull request:
**Merge checklist:**
Check off if complete *or* not applicable:
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Fixup commits are squashed away
- [x] Unit tests added/updated
- [x] Manual testing instructions provided
- [x] Noted any: Concerns, dependencies, migration issues, deadlines, tickets

## Testing instructions
1. Use the `master` branch of the `edx-platform`.
2. Clone and install this package. Run migrations.
3. Upload the assets attached to the Jira ticket (private).
4. [Create](localhost:18000/admin/openedx_certificates/externalcertificatetype/add) two certificate types:
   1. Completion:
      1.  Retrieval func: `retrieve_course_completions`
      2. Generation func: `generate_pdf_certificate`
      3. Custom options:
           ```json
            {
                "name_y": 290,
                "course_name_y": 210,
                "template": "template_completion",
                "template_two_lines": "template_completion_two_line",
                "required_completion": 0.9
            }
            ```
   2. Achievement:
      1.  Retrieval func: `retrieve_subsection_grades`
      2. Generation func: `generate_pdf_certificate`
      3. Custom options:
           ```json
            {
                "name_y": 290,
                "course_name_y": 220,
                "template": "template_achievement",
                "required_grades": {
                    "Homework": 0.4,
                    "Exam": 0.9,
                    "Total": 0.8
                }
            }
            ```
5. [Create](localhost:18000/admin/openedx_certificates/externalcertificatecourseconfiguration/add) two course certificate configurations:
   1. Completion - you can use `course-v1:edX+DemoX+Demo_Course` as the course ID and override the `required_completion` to make the testing easier.
   2. Achievement - you can use `course-v1:edX+DemoX+Demo_Course` as the course ID and override the `required_grades` to make the testing easier.
6. Press the "generate certificates" button for each of the created course configurations.
7. [Check](http://localhost:18000/admin/openedx_certificates/externalcertificate/) that the certificates were generated correctly.

## Notes
1. The docs are deployed to https://openedx-certificates.readthedocs.io.
2. The pluggability of processors and generators will be included in the next iteration.

*Private-ref:* [BB-7595](https://tasks.opencraft.com/browse/BB-7595)

Agrendalath commented on docs/decisions/0002-architecture.rst:
```diff
@@ -0,0 +1,119 @@
+Decision
+********
+
+.. graphviz::
```
It does not seem to be working on ReadTheDocs. When I build it locally, it looks like this:

![image](https://github.com/open-craft/openedx-certificates/assets/3189670/3ad665c9-5537-4edb-81f1-f63aa873eb30)

Agrendalath replied:
Fixed in c9fb3e96d89606dec45350c36c50a85ded8b506e.

bradenmacdonald commented on openedx_certificates/compat.py:
```diff
@@ -0,0 +1,76 @@
+
+@contextmanager
+def prefetch_course_grades(course_id: CourseKey, users: list[User]):
+    """Prefetch the course grades from Open edX."""
```
What does "prefetch" mean? How is it different from "fetching" the grades?

bradenmacdonald commented on openedx_certificates/generators.py:
```diff
@@ -0,0 +1,153 @@
+The functions prefixed with `generate_` are automatically detected by the admin page and are used to generate the
+certificates for the users.
+
+We will move this module to an external repository (a plugin).
```
Making it a plugin and putting it in a separate python package is fine, but moving it to a separate repo too early will only slow down development imho.

bradenmacdonald commented on openedx_certificates/generators.py:
```diff
@@ -0,0 +1,153 @@
+    return url
+
+
+def generate_pdf_certificate(course_id: CourseKey, user: User, certificate_uuid: UUID, options: dict[str, Any]) -> str:
```
Will every `generate_` method use this same signature? (for badges, external certs, other types...?)

It seems to be missing the date at which the cert was earned, which may be different from the date when the PDF is generated.

bradenmacdonald commented on openedx_certificates/generators.py:
```diff
@@ -0,0 +1,153 @@
+        default_storage.delete(output_path)
+    default_storage.save(output_path, certificate_file)
+    if isinstance(default_storage, FileSystemStorage):
+        url = f"{settings.LMS_ROOT_URL}{settings.MEDIA_URL}{output_path}"
```
The "URL to the default storage" should probably be defined in `compat.py`, not directly referencing `LMS_ROOT_URL` and `MEDIA_URL` here ?

bradenmacdonald commented on openedx_certificates/models.py:
```diff
@@ -1,5 +1,345 @@
+            certificate.status = ExternalCertificate.Status.GENERATING
+            certificate.generation_task_id = celery_task_id
+            certificate.save()
+        except ExternalCertificate.DoesNotExist:
```
Nit: Seems like you could use [`update_or_create`](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#update-or-create) here to shorten this code.

bradenmacdonald commented on openedx_certificates/models.py:
```diff
@@ -1,5 +1,345 @@
+            try:
+                module_path, func_name = func_path.rsplit('.', 1)
+                module = import_module(module_path)
+                getattr(module, func_name)  # Will raise AttributeError if the function does not exist.
```
Nit: I think it would be nice if the above three lines to load the actual function (for `retrieval_func` and `generation_func`) were implemented as properties or methods on this class. They are currently duplicated in `get_eligible_user_ids` and `generate_certificate_for_user` for example.

bradenmacdonald commented on openedx_certificates/models.py:
```diff
@@ -1,5 +1,345 @@
+            generation_func = getattr(generation_module, generation_func_name)
+
+            # Run the functions. We do not validate them here, as they are validated in the model's clean() method.
+            certificate.download_url = generation_func(self.course_id, user, certificate.uuid, custom_options)
```
Is `download_url` the most generic data to receive back from a generation function? I'm thinking of badges which are more displayed then downloaded. Also external credentials like something on Accredible aren't downloaded either. Maybe just calling it `detail_url` will work though because I assume everything has a URL to view it in detail.

The most general approach would be to let `generation_func` return a dict of JSON data, which in the case of PDFs includes a `download_url` key.

I'm also wondering about how the URLs work for PDF certificates. I assume we want to guarantee the URLs are stable forever, so using the underlying django storage URL of the PDF may not be ideal. Is there some intermediate layer that provides permalinks for each certificate and will redirect or serve the PDF as needed?

bradenmacdonald commented on openedx_certificates/models.py:
```diff
@@ -1,5 +1,345 @@
+    This model contains information about the related course, the user who earned the certificate,
+    the download URL for the certificate PDF, and the associated certificate generation task.
+
+    .. note:: The ID field is not a conventional auto-incrementing integer, but a value
+       that allows for old certificates with custom IDs.
```
```suggestion
    .. note:: Certificates are identified by UUIDs rather than integer keys. Not only
       does this prevent enumeration attacks / privacy leaks, but it is backwards-
       compatible with old certificate formats, so we can import and represent them
       here without changing their IDs.
```
I was slightly confused that there is no `ID` field here, because it seems pretty obvious that a field called `uuid` is not going to be an integer, so I thought maybe you were talking about some other field. Or are you talking about `legacy_id`?

bradenmacdonald commented on openedx_certificates/models.py:
```diff
@@ -1,5 +1,345 @@
+        help_text=_('Status of the certificate generation task'),
+    )
+    download_url = models.URLField(blank=True, help_text=_('URL of the generated certificate PDF (e.g., to S3)'))
+    legacy_id = models.IntegerField(null=True, help_text=_('Legacy ID of the certificate imported from another system'))
```
Probably needs `blank=True`.

bradenmacdonald commented on openedx_certificates/models.py:
```diff
@@ -1,5 +1,345 @@
+    )
+    user_id = models.IntegerField(help_text=_('ID of the user receiving the certificate'))
+    user_full_name = models.CharField(max_length=255, help_text=_('User receiving the certificate'))
+    course_id = CourseKeyField(max_length=255, help_text=_('ID of a course for which the certificate was issued'))
```
Why do we store the user's full name in a separate field, but we don't store the course name in a separate field?

bradenmacdonald commented on openedx_certificates/models.py:
```diff
@@ -1,5 +1,345 @@
+        GENERATING = 'generating', _('Generating')
+        AVAILABLE = 'available', _('Available')
+        ERROR = 'error', _('Error')
+        INVALIDATED = 'invalidated', _('Invalidated')
```
Can you document what `INVALIDATED` means, with a comment?

bradenmacdonald commented on openedx_certificates/pipelines.py:
```diff
@@ -0,0 +1,18 @@
+"""TODO: Add some docstring here. We may also want to move this to a different file."""
```
docstring

bradenmacdonald commented on openedx_certificates/processors.py:
```diff
@@ -0,0 +1,224 @@
+"""
+This module contains processors for certificate criteria.
+
+The functions prefixed with `retrieve_` are automatically detected by the admin page and are used to retrieve the
```
> by the admin page

I would have thought it's a cron job calling these, not manual trigger from the admin page?

bradenmacdonald commented on openedx_certificates/compat.py:
```diff
@@ -0,0 +1,76 @@
+
+
+def get_course_grade_factory():  # noqa: ANN201
+    """Get the course grade factory from Open edX."""
```
Can you mention what this factory does, and how it interacts with `prefetch_...`? I thought maybe it was for mocking grades during testing, but it seems that's not the case.

bradenmacdonald commented on openedx_certificates/processors.py:
```diff
@@ -0,0 +1,224 @@
+The functions prefixed with `retrieve_` are automatically detected by the admin page and are used to retrieve the
+IDs of the users that meet the criteria for the certificate type.
+
+We will move this module to an external repository (a plugin).
```
I would suggest just putting this into a `plugins/completion/processor.py` file in the same repo, along with the corresponding generators (for "Certificate of completion" & "Badge of achievement/completion")

bradenmacdonald commented on requirements/quality.in:
```diff
@@ -4,9 +4,6 @@
 
 -r test.txt               # Core and testing dependencies for this package
 
```
Nit: Personally I don't think it's better to have so many different requirements files. Just three - `base`, `dev`, and `constraints` - would cover every use case, even if `dev` will sometimes install more than is needed for a specific use case like `quality` or `ci`. But no need to change anything, just feel like mentioning it.

bradenmacdonald commented on setup.py:
```diff
@@ -1,31 +1,33 @@
-"""
-Package metadata for openedx_certificates.
-"""
+"""Package metadata for openedx_certificates."""
```
Do we still need `setup.py`? Can't we declare all this using `pyproject.toml` which is the replacement, and which is more efficient since it doesn't have to be run as code before it can be used?

bradenmacdonald reviewed (commented):
I know this is a lot of feedback but I do think this looks nice! Feel free to push back on anything; a lot of this is trying to understand and push for clear documentation.

tecoholic commented on docs/decisions/0002-architecture.rst:
```diff
@@ -0,0 +1,119 @@
+            // Processes
+            edge [fontcolor=black, color=red]
+            Schedule -> Processing [label="trigger"]
+            Processing -> Generation [label="provide elgible users"]
```
Nit:
```suggestion
            Processing -> Generation [label="provide eligible users"]
```

Agrendalath commented:
Update: I addressed the comments that could have been addressed within the available budget. I created a GH issue and TODOs for most of the remaining ones. I'm merging this to unblock other tickets.
